Quality of the paired end reads was assessed using FastQC and reads were quality filtered and screened for adapters using Trimgalore and reads were then aligned to the {{ snakemake.config["genome"] }} reference genome using bowtie with the parameters :code:`--very-sensitive --maxins {{ snakemake.config["bowtie2"]["maxins"] }} --no-mixed --no-discordant `. PCR duplicate reads were removed using Picard MarkDuplicates.  Peaks were called subsequently using macs2 with the parameters :code:`--keep-dup all -f BAMPE -q {{ snakemake.config["macs2"]["qvalue"] }}`, while simultaneously generating bedgraph files in Reads Per Million scale. Peaks were filtered against the ENCODE blacklist regions. The bedgraph files were converted to bigWig file format using the bedGraphToBigWig utility.
